### PR TITLE
[GPE] Split critical edges before marking blocks

### PR DESCRIPTION
**Pull request by rxwei**
Bug [SR-8373](https://bugs.swift.org/browse/SR-8373) originates from the [Autoencoder](https://github.com/tensorflow/swift-models/tree/master/Autoencoder) model.

This patch fixes it by splitting `cond_br` critical edges before marking blocks.

After splitting critical edges, the compiler no longer sees `infLoop2` as a self loop, so it can be correctly lowered.

### CODE REVIEW

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by mhong** [comment]
Looks good! This may have also fixed https://bugs.swift.org/browse/SR-8241

File: lib/SILOptimizer/Mandatory/TFPartition.cpp [hunk 1]
```diff
@@ -622,6 +624,7 @@ class TFFunctionPartition {
   DenseMap<StringRef, std::unique_ptr<LoweredGraphFunction>> &graphFunctions;
   GraphFunctionDeviceInfo deviceInfo; // Device placement info.
   DominanceInfo &DI;
+  SILLoopInfo &LI;
```
Comment (lattner) [inline lib/SILOptimizer/Mandatory/TFPartition.cpp]: This shouldn't need LoopInfo.  You can just pass a null pointer into splitAllCondBrCriticalEdgesWithNonTrivialArgs

File: lib/SILOptimizer/Mandatory/TFPartition.cpp [hunk 2]
```diff
@@ -1963,6 +1967,9 @@ bool TFFunctionPartition::markFunction(bool &hasTensorOps) {
       loggedInput = true;
     }
   };
+  
+  // Split all critical edges for block marking.
+  splitAllCondBrCriticalEdgesWithNonTrivialArgs(hostFn, &DI, &LI);
```
Comment (lattner) [inline lib/SILOptimizer/Mandatory/TFPartition.cpp]: Please add a fixme to remove this when TFPartition is integrated into deabstraction, thanks!

File: lib/SILOptimizer/Mandatory/TFPartition.cpp (hunk 1 above)
Comment (rxwei) [reply lib/SILOptimizer/Mandatory/TFPartition.cpp]: Addressed in https://github.com/apple/swift/pull/18437.

File: lib/SILOptimizer/Mandatory/TFPartition.cpp (hunk 2 above)
Comment (rxwei) [reply lib/SILOptimizer/Mandatory/TFPartition.cpp]: Addressed in #18437.

**Pull request merged**